### ISSUE TITLE
PSCommandNotFoundSuggestion is suggesting commands when it should not

### ISSUE THREAD

**Issue by Jaykul** [PowerShell/PowerShell#10213]
# Steps to reproduce

If you use `Get-Command -ErrorAction Ignore` to test for the existence of a command, you get "Suggestion" output which is _confusing_ to the end user and seems to indicate an error, when in fact there is no problem.

For example, in one script I do this:
```powershell
$Version = if (Get-Command gitversion -ErrorAction SilentlyContinue) { 
    gitversion -showvariable semver
} else { 
   "1.0.0-rc"
}
```

# Expected behavior

When `Get-Command` is called and the error is explicitly suppressed, there should be no suggestions printed to the console.

Frankly, it might be better if `Get-Command` doesn't print suggestions at all, or perhaps only when it's being called interactively from the prompt (and the error isn't explicitly suppressed).

```none

```

# Actual behavior

I get a completely off-the-wall suggestion from the PSDeploy module:

```none
Suggestion [4,General]: The most similar commands are: WithOptions
```

Frankly, that command is not at all similar, in my human brain, and it really should not be mentioned at all. I can't understand how that's past the similarity threshold.

# Environment data

```none
Name                           Value
----                           -----
PSVersion                      6.2.1
PSEdition                      Core
GitCommitId                    6.2.1
OS                             Microsoft Windows 10.0.18362
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

**Comment by Jaykul** [comment]
The suggestions need some tuning. Right now (without looking at the code) it seems to be based on levenshtein edit distance, but the _threshold isn't being tuned based on how short the command_ is, so if you miss-type a _short_ command, you get a **lot** of suggestions, most of which have no similarity except length.

```posh
> Get-Command abc -ea 0                                                                                                                                                                 
Suggestion [4,General]: The most similar commands are: ac, clc, dbp, ebp, gbp, gc, rbp, sbp, fc, c.
```

```posh
> Get-Command zzz -ea 0                                                                                                                                                                                                                                                                                                                                             Suggestion [4,General]: The most similar commands are: gtz, stz, z, %, ?, cp.
```

**Comment by bergmeister** [comment]
Yes, I agree that this definitely needs tweaking.
But I've always also found it annoying that `Get-Command` returns an error if nothing was found (yes, I had to write similar code to yours before as well....). It would be a breaking change but in an ideal world the command should just return nothing and the caller can then write conditional logic 
What annoy me about Get-Command as well is that it loads the module into memory by doing a full Import-Module (which makes this check very expensive). For a binary module this means I am stuck with that loaded version since PS cannot unload it (maybe that is possible with recent additions in .Net Core?).

In order to avoid breaking changes, what about a more lightweight and usable Test-Command cmdlet?

**Comment by lzybkr** [comment]
@Jaykul - Your description mentions using `-ErrorAction Ignore`, but your sample uses `-ErrorAction SilentlyContinue`. That might explain what's going on.

@bergmeister - You can use a wildcard to both avoid an error and avoid loading the module. It does mean you have to check the result(s) to make sure you didn't pick up unexpected results though.

**Comment by sba923** [comment]
That could be a different issue, but sometimes (don't know exactly under what conditions) when my `profile.ps1` runs, and executes code such as

````
if (running_at_home)
{
...
}
````

where `running_at_home` resolves to `s:\bin\home\running_at_home.ps1`

I get:

````
Suggestion [4,General]: The most similar commands are: running_at_home.ps1
````

**Comment by sba923** [comment]
FWIW: the issue is still there with 7.1 preview 3 (which is probably expected since it's still open 😉)

**Comment by vexx32** [comment]
Yeah. The reality is that the suggestions system needs to be made more robust and only be used when an error record is actually being displayed.

I have some past work sorting the majority of that out if anyone would like to pick it up. It needs (at minimum) the entirety of the ErrorRecord format code rewritten, since @SteveL-MSFT made significant changes to it as well in order to implement ConciseView, and my original code was also a complete rewrite of the formatter code as well.

Steve has expressed interest in completing that work, but if anyone else is interested as well, the PR linked above has the majority of the code sorted out; you'll just need to rewrite my additions errorrecord formatting code (probably from scratch), and I think there were also a couple minor changes needed to the rest of the code to get everything sorted out neatly.

**Comment by sba923** [comment]
Thanks @vexx32 for the clarification. Can you point me at the location in the code where all this is handled? I might want to take a look...

**Comment by vexx32** [comment]
See my changes for #9115.

Basically the approach I think will be most useful is:

- Completely remove existing suggestion API. It's inaccessible and host-specific so folks using custom hosts cannot get any benefit from it.
- Add a property to Error record to hold suggestion(s).
- Modify ErrorRecord formatter to show suggestions (in a different color to the usual error color display).

I think the only complication I ran into was that CommandNotFoundException is a bit weird under the hood and if you're not careful you can potentially end up causing an infinite loop, though I think the code I have in my PR does successfully avoid it.

**Comment by SteveL-MSFT** [comment]
This should now be fixed in 7.4-preview.1 based on the FeedbackProvider which works differently than the old SuggestionFramework
